### DISCUSSION TITLE
[lodash] `isObject` type guard isn't entirely accurate

### DISCUSSION THREAD

**Discussion by thomasttvo** [DefinitelyTyped/DefinitelyTyped#59138] [Issues with a @types package]
Instead of `is object`, it should be `is Record<string | number | symbol, unknown>`. This better illustrates the nature of the input value should the result be `true`.

Why this is better?

```ts
  // With `is object`
 if(isObject(a) && a.prop === undefined) // get TS error "'prop' does not exist"

 // With `is Record<string | number | symbol, unknown>`
 if(isObject(a) && a.prop === undefined) // no error, type of `prop` is `unknown`, which is how it should be.
```

[a comment by an automated account is not shown]

**Comment by aj-r** [comment]
I'd like to avoid extra complexity unless it's really necessary. We recently got dinged by making a type more complex which made it "better" but also broke other use cases.

In this case I'm not sure that `Record<string | number | symbol, unknown>` adds any value. I think your issue only comes up if `a` has type  `unknown`. Even then, I think you should be able to do something like:

```
 if(isObject(a) && 'prop' in a) // no error
```

Alternatively, If you give `a` a more specific type then this should not be an issue, e.g.

```
let a: Record<string | number | symbol, unknown> | undefined;
```

If neither of those work, you could always do a cast somewhere. I think if `a` is `unknown` then it's reasonable to need to cast at some point.

**Reply by thomasttvo** [reply]
I don't think the first option would work.

```ts
const a: object = {};

if('prop' in a) {
  const x = a.prop; // TS error: 'prop' does not exist on type 'object'
}
```

In my case, `a` is `unknown`, and it's not safe to cast an `unknown`. The best practice is to use typeguard and we can't create a typeguard for every single case, which is why we need `isObject` to act as a better typeguard.
